### PR TITLE
Allow to config MAUTH_CACHE_TTL via an ENV var, now defaults to 10 minutes

### PR DESCRIPTION
**Pull request by mjobin-mdsol**

### CODE REVIEW

File: lib/mauth/client/security_token_cacher.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@ module LocalAuthenticator
       class SecurityTokenCacher
 
         class ExpirableSecurityToken < Struct.new(:security_token, :create_time)
-          CACHE_LIFE = 60
+          CACHE_LIFE = ENV['MAUTH_CACHE_TTL'].to_i >= 60 ? ENV['MAUTH_CACHE_TTL'].to_i : 600
```
Comment (jfeltesse-mdsol) [inline lib/mauth/client/security_token_cacher.rb:7]: this does more than simply allowing to read the value from env, it would read it only if it's set and >= 60, which is misleading to me. What if I want to pass 30?

This should check the value of the var is present and is a positive int, falling back to 600.

**Review by jfeltesse-mdsol — COMMENTED** [review]
code change = changelog, version bump needed.

File: lib/mauth/client/security_token_cacher.rb (hunk 1 above)
Comment (mjobin-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: I don't think we should allow it to be lower, it does not make sense. The purpose is to bump it.
This value was set in 2012 or earlier and was never reviewed.

Ideally, it should take the cache headers from the mauth service and respect that, but that's above the scope for now.

Comment (mjobin-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: I believe I should keep the same default value though, but I wanted to change it right away for easy test deploy.

Comment (jfeltesse-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: > I don't think we should allow it to be lower, it does not make sense. The purpose is to bump it.

IMHO this is not the job of this library to decide what makes sense or not, I can imagine for instance for testing purposes you may be happy to be able to set it at a very low value.

If you really want to disallow lower values then the logic should be fleshed out more, for instance detecting those and output some message like "mauth cache TTL must be at least 10 mins" to tell the user about it.
Nothing worse than a library that undercuts the user's config without saying so.

And it also makes me think, to be complete, that PR should document that config and also the config files + parser should support it too.

Comment (mjobin-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: I am not making 10 minutes a minimum. I want to test deploy this before making it complete.

60s, the hardcoded value is a good minimum. and should probably remain default.

would anyone want to decrease that? they should bring in their use case. lets try to focus on use cases at hand.

original hardcoded value 2011
https://github.com/mdsol/rack-mauth/blob/0befb8b724fcf2b7b7dbd4365242024ca7328b8b/lib/rack/mauth.rb#L55

refactored and moved to this gem in 2012
https://github.com/mdsol/mauth-client-ruby/commit/d9b12d704c37ee1036d6d52c15b665bec9e7b020#

Comment (jfeltesse-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: Please re-read my comment. I'm not arguing what's a good default value.

I'm saying the logic is misleading, wherein you're allowing the value to be passed from env but if it's below the minimum that's hardcoded, the library is going to silently override the value with no error, no logging, nothing. That's just not good UX.

Comment (jcarres-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: I guess this is a quick fix but the real logic we want is to do whatever the mauth server is telling us to do, of if it does not say anything, only then have a default. 
Dotnet does it so:
https://github.com/mdsol/mauth-client-dotnet/blob/87ba13240530018651cb3a419dfa9d99da0d4e05/src/Medidata.MAuth.Core/MAuthAuthenticator.cs#L145

Comment (ejinotti-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: FWIW MAuth service does have a config var `TOKEN_EXPIRATION` for controlling the max age in its response header which defaults to `60` when not given.  Unfortunately, this var is not actually listed in the app description file to we couldn't actually use it without a release 😬

Comment (ejinotti-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: https://github.com/mdsol/mauth/pull/249

Comment (jcarres-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: Why so low (60s) ? It seems to me mauth keys do not change ever

Comment (ejinotti-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: Not really sure why we went with `60` as a default.  Here is the PR that added caching headers to the response:

https://github.com/mdsol/mauth/pull/188

Before that PR is was either not putting cache-control headers or always putting the "no-cache" value.

I think maybe we went with `60` to just match what `mauth-client-ruby` does, but perhaps @ejennings-mdsol remembers more?  The main objective of that PR was just to get .NET clients to cache the tokens as they were looking for the header and thus not doing any caching, whereas the ruby client doesn't care about the header and just does its own thing.

I agree that the best approach would be to fix this client to respect the header.  Also maybe should do another PR to MAuth service to make the default something more sensible that `60` if everyone agrees.

Comment (cmcinnes-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: You could argue that security tokens should not be cached too long in mauth clients in case a private key has been compromised and a security token needs to be deleted. if we cached them for a day an attacker could have free reign to our apis for up to a day even after we delete the token in mauth service. there are ways around this (config changes, redeploy) but would be annoying.
I'm not arguing against raising the default ttl, just don't think we should go crazy high. i think (hope) the risk of a private key being leaked is low.

**Comment by cmcinnes-mdsol** [comment]
thanks @mjobin-mdsol ! i do agree with others that this configuration option should be documented: https://github.com/mdsol/mauth-client-ruby#configuration and perhaps here https://github.com/mdsol/mauth-client-ruby/blob/master/doc/mauth.yml.md#configuration-options

File: spec/support/shared_examples/authenticator_base.rb [hunk 2]
```diff
@@ -8,7 +8,7 @@
     end
 
     it "considers an authentically-signed request to be inauthentic when it's too old or too far in the future" do
-      [-301, 301].each do |time_offset|
+      [-301, 601].each do |time_offset|
```
Comment (ejennings-mdsol) [inline spec/support/shared_examples/authenticator_base.rb:11]: Why was this changed?  5 minutes is still the `ALLOWED_DRIFT_SECONDS` value.

It would probably be clearer if this were:
```
drift = MAuth::Client::AuthenticatorBase::ALLOWED_DRIFT_SECONDS
[-(drift + 1),  drift + 1].each do |time_offset|
```

File: lib/mauth/client/security_token_cacher.rb (hunk 1 above)
Comment (ejennings-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: Sorry @ejinotti-mdsol, I don't remember.

I agree on respecting the mauth service header, which could be out of scope for this PR, but what is the business driver for this PR?

Comment (ykitamura-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: .NET and python clients are respecting the mauth service header: 
- https://github.com/mdsol/mauth-client-dotnet/tree/release/v5.1.1#does-medidatamauth-support-caching
- https://github.com/mdsol/mauth-client-python/blob/1.2.1/mauth_client/key_holder.py#L37

Comment (ejinotti-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: @ejennings-mdsol I think the driver was that Jordi and Mathieu noticed a lot of MAuth calls for security tokens from certain apps, Coder and Dalton being the worst offenders:

image

I guess since `mauth-client` caches in local memory and Dalton can run up to 25 containers with up to 8 passenger processes each that can be up to 60 x 25 x 8 = 12,000 MAuth calls per hour for each security token.

Comment (jcarres-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: @ejennings-mdsol We noticed the high traffic between Coder and Mauth, turns out they are using a very old dotnet mauth client with no caching. I think here Mathieu is trying to improve performance on the Ruby side also. 
We have also. noticed that while average call to mauth server is very fast, the 95th percentile is pretty bad.

File: spec/support/shared_examples/authenticator_base.rb (hunk 2 above)
Comment (mjobin-mdsol) [reply spec/support/shared_examples/authenticator_base.rb:11]: Honestly, I'm not sure why this needed to be bumped.

File: lib/mauth/client/security_token_cacher.rb (hunk 1 above)
Comment (mjobin-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: I thought the first step was to test deploy with a new value.
Then gather comments on what are the risk to bump this.
I don't have enough knowledge on to really judge properly.
@cmcinnes-mdsol correct me if I am wrong.
but if the client cache the public key for an day or a week, and suddenly the private key has been compromised.
the server will be redeployed with a new key and the client will stop working until they refresh their cache?

I think we all agree the client should be watching for cache-control headers and respecting that.
removing the need of this ENV var here, or the hardcoded value.

and ideally, we should bump the expiry on the mauth server side as high as possible.

Comment (ykitamura-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: how about utilizing [faraday-http-cache](https://github.com/sourcelevel/faraday-http-cache) and `Rails.cache` to cache in Redis?

Comment (mjobin-mdsol) [reply lib/mauth/client/security_token_cacher.rb:7]: @ykitamura-mdsol that sounds like a good suggestion

**Comment by mjobin-mdsol** [comment]
replaced by https://github.com/mdsol/mauth-client-ruby/pull/50
